Library/Homebrew/cmd/commands.rb

```diff
@@ -8,51 +8,70 @@ module Homebrew
   module_function
 
   def commands
-    if ARGV.include? "--quiet"
-      cmds = internal_commands + external_commands
-      cmds += internal_developer_commands
-      cmds += HOMEBREW_INTERNAL_COMMAND_ALIASES.keys if ARGV.include? "--include-aliases"
-      puts Formatter.columns(cmds.sort)
-    else
-      # Find commands in Homebrew/cmd
-      puts "Built-in commands"
-      puts Formatter.columns(internal_commands)
-
-      # Find commands in Homebrew/dev-cmd
-      puts
-      puts "Built-in developer commands"
-      puts Formatter.columns(internal_developer_commands)
-
-      # Find commands in the path
-      unless (exts = external_commands).empty?
-        puts
-        puts "External commands"
-        puts Formatter.columns(exts)
+    # TODO: Goal is to not require this function at all.
+    command = CommandsCommand.new
+    command.check_for_errors
+    command.commands
+  end
+
+  class CommandsCommand < Command
+    def initialize
+      options do
+        cmd_name "commands"
+        desc "Show a list of built-in and external commands."
+        option "quiet", desc: "list only the names of commands without the header." do
+          option "include-aliases", desc: "the aliases of internal commands will be included."
```

Comment:
@woodruffw, @MikeMcQuaid repeatedly said that we should deviate from the current `--help` and `manpages` output as little as possible (e.g https://github.com/Homebrew/brew/pull/2420#discussion_r117177652). The current implementation does exactly that. Currently it generates the exact same `help` and `man` output for many commands. What you are suggesting will completely change the `--help` and `manpage` output for each command. 

First case:
```
option "foo", desc: "do foo" do
  option "bar", desc: "do bar" do
    option "baz", desc: "do baz" do

    end
  end
end
```

Second case:
```
option "foo", desc: "do foo"
option "bar", desc: "do bar"
option "baz", desc: "do baz"
```

So currently, the `--help` output for first case will be:
```
brew commands [--foo [--bar [--baz]]]

    If --foo is passed, do foo
    With --bar, do bar
    With --baz, do baz
```

And the `--help` output for second case will be:
```
brew commands [--foo] [--bar] [--baz]

    If --foo is passed, do foo
    
    If --bar is passed, do bar

    If --baz is passed, do baz
```